peps/pep-0740.rst

```diff
@@ -205,110 +181,340 @@ changes to it:
+  asynchronously or only appear after the file is already public on the index,
+  requiring the index to insert them into the provenance *post facto*.
 
-These changes require a version change to the JSON API:
+* Addition of signing identities: when new attestations are added to the
+  provenance object, their verification may require new signing identities
+  (such as public keys or Trusted Publisher identities).
+
+.. _attestation-verification:
+
+Attestation verification
+------------------------
+
+Verifying an attestation object requires verification of each of the following:
 
-* The ``api-version`` must specify version 1.2 or later.
+* ``version`` is ``1``. The verifier **MUST** reject any other version.
+* ``verification_material.certificate`` is a valid signing certificate, as
+  issued by an *a priori* trusted authority (such as a root of trust already
+  present within the verifying client).
+* ``verification_material.certificate`` identifies an appropriate signing
+  subject, such as the machine identity of the Trusted Publisher that published
+  the package.
+* ``message_signature`` can be verified by ``verification_material.certificate``,
+  using the reconstructed attestation payload as the cleartext input. The
+  verifier **MUST** reconstruct the attestation payload itself.
+
+In addition to the above required steps, a verifier **MAY** additionally verify
+``verification_material.transparency_entries`` on a policy basis, e.g. requiring
+at least one transparency log entry or a threshold of entries. When verifying
+transparency entries, the verifier **MUST** confirm that the inclusion time for
+each entry lies within the signing certificate's validity period.
 
 Security Implications
 =====================
 
-This PEP is "mechanical" in nature; it provides only the plumbing for future
-digital attestations on package indices, without specifying their concrete
-cryptographic details.
+This PEP is primarily "mechanical" in nature; it provides layouts for
+structuring and serving verifiable digital attestations without specifying
+higher level security "policies" around attestation validity, thresholds
+between attestations, and so forth.
+
+Cryptographic agility in attestations
+-------------------------------------
+
+Algorithmic agility is a common source of exploitable vulnerabilities
+in cryptographic schemes. This PEP limits algorithmic agility in two ways:
 
-As such, we do not identify any positive or negative security implications
-for this PEP.
+* All algorithms are specified in a single suite, rather than a geometric
+  collection of parameters. This makes it impossible (for example) for an
+  attacker to select a strong signature algorithm with a weak hash function,
+  compromising the scheme as a whole.
+* Attestation objects are versioned, and may only contain the algorithmic
+  suite or suites specified for their version. If a specific suite
```

Comment:
Above, in the `Attestation objects` section, we say:
> Each version is tied to a **single** cryptographic suite to minimize unnecessary cryptographic agility.

So here we should remove the `or suites`,  or change the above to say that a version can be tied to multiple suites.
```suggestion
  suite specified for their version. If a specific suite
```

Reply:
Yes, thanks!